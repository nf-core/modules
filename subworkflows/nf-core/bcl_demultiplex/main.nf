#!/usr/bin/env nextflow

//
// Demultiplex Illumina BCL data using bcl-convert or bcl2fastq
//

include { BCLCONVERT } from "../../../modules/nf-core/bclconvert/main"
include { BCL2FASTQ  } from "../../../modules/nf-core/bcl2fastq/main"

workflow BCL_DEMULTIPLEX {
    take:
        ch_flowcell     // [[id:"", lane:""],samplesheet.csv, path/to/bcl/files]
        demultiplexer   // bclconvert or bcl2fastq

    main:
        ch_versions = Channel.empty()
        ch_fastq    = Channel.empty()
        ch_reports  = Channel.empty()
        ch_stats    = Channel.empty()
        ch_interop  = Channel.empty()

        // Split flowcells into separate channels containg run as tar and run as path
        // https://nextflow.slack.com/archives/C02T98A23U7/p1650963988498929
        ch_flowcell
            .branch { meta, samplesheet, run ->
                tar: run.toString().endsWith(".tar.gz")
                dir: true
            }.set { ch_flowcells }

        ch_flowcells.tar
            .multiMap { meta, samplesheet, run ->
                samplesheets: [ meta, samplesheet ]
                run_dirs: [ meta, run ]
            }.set { ch_flowcells_tar }

        // Runs when run_dir is a tar archive
        // Re-join the metadata and the untarred run directory with the samplesheet
        ch_flowcells_tar_merged = ch_flowcells_tar.samplesheets.join( ch_flowcells_tar.run_dirs )

        // Merge the two channels back together
        ch_flowcells = ch_flowcells.dir.mix(ch_flowcells_tar_merged)

        // MODULE: bclconvert
        // Demultiplex the bcl files
        if (demultiplexer == "bclconvert") {
            BCLCONVERT( ch_flowcells )
            ch_fastq    = ch_fastq.mix(BCLCONVERT.out.fastq)
            ch_interop  = ch_interop.mix(BCLCONVERT.out.interop)
            ch_reports  = ch_reports.mix(BCLCONVERT.out.reports)
            ch_versions = ch_versions.mix(BCLCONVERT.out.versions)
        }

        // MODULE: bcl2fastq
        // Demultiplex the bcl files
        if (demultiplexer == "bcl2fastq") {
            BCL2FASTQ( ch_flowcells )
            ch_fastq    = ch_fastq.mix(BCL2FASTQ.out.fastq)
            ch_interop  = ch_interop.mix(BCL2FASTQ.out.interop)
            ch_reports  = ch_reports.mix(BCL2FASTQ.out.reports)
            ch_stats    = ch_stats.mix(BCL2FASTQ.out.stats)
            ch_versions = ch_versions.mix(BCL2FASTQ.out.versions)
        }

        // Generate meta for each fastq
        ch_fastq_with_meta = generate_fastq_meta(ch_fastq)

    emit:
        fastq    = ch_fastq_with_meta
        reports  = ch_reports
        stats    = ch_stats
        interop  = ch_interop
        versions = ch_versions
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    FUNCTIONS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

// Add meta values to fastq channel
def generate_fastq_meta(ch_reads) {
    // Create a tuple with the meta.id and the fastq
    ch_reads.transpose().map{
        fc_meta, fastq ->
        def meta = [
            "id": fastq.getSimpleName().toString() - ~/_R[0-9]_001.*$/,
            "samplename": fastq.getSimpleName().toString() - ~/_S[0-9]+.*$/,
            "readgroup": [:],
            "fcid": fc_meta.id,
            "lane": fc_meta.lane
        ]
        meta.readgroup = readgroup_from_fastq(fastq)
        meta.readgroup.SM = meta.samplename

        return [ meta , fastq ]
    }
    // Group by meta.id for PE samples
    .groupTuple(by: [0])
    // Add meta.single_end
    .map {
        meta, fastq ->
        if (fastq.size() == 1){
            meta.single_end = true
        } else {
            meta.single_end = false
        }
        return [ meta, fastq.flatten() ]
    }
}

// https://github.com/nf-core/sarek/blob/7ba61bde8e4f3b1932118993c766ed33b5da465e/workflows/sarek.nf#L1014-L1040
def readgroup_from_fastq(path) {
    // expected format:
    // xx:yy:FLOWCELLID:LANE:... (seven fields)

    def line

    path.withInputStream {
        InputStream gzipStream = new java.util.zip.GZIPInputStream(it)
        Reader decoder = new InputStreamReader(gzipStream, 'ASCII')
        BufferedReader buffered = new BufferedReader(decoder)
        line = buffered.readLine()
    }
    assert line.startsWith('@')
    line = line.substring(1)
    def fields = line.split(':')
    def rg = [:]

    // CASAVA 1.8+ format, from  https://support.illumina.com/help/BaseSpace_OLH_009008/Content/Source/Informatics/BS/FileFormat_FASTQ-files_swBS.htm
    // "@<instrument>:<run number>:<flowcell ID>:<lane>:<tile>:<x-pos>:<y-pos>:<UMI> <read>:<is filtered>:<control number>:<index>"
    sequencer_serial = fields[0]
    run_nubmer       = fields[1]
    fcid             = fields[2]
    lane             = fields[3]
    index            = fields[-1] =~ /[GATC+-]/ ? fields[-1] : ""

    rg.ID = [fcid,lane].join(".")
    rg.PU = [fcid, lane, index].findAll().join(".")
    rg.PL = "ILLUMINA"

    return rg
}
