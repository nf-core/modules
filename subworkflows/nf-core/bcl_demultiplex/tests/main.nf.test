// nf-core subworkflows test bcl_demultiplex
nextflow_workflow {

    name "Test Subworkflow test_bcl_demultiplex_bclconvert"
    script "../main.nf"
    config "./nextflow.config"
    workflow "BCL_DEMULTIPLEX"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/bcl_demultiplex"
    tag "bclconvert"
    tag "bcl2fastq"

    test("bclconvert") {

        when {

            workflow {
                """
                input[0] = Channel.value([
                    [id:'test', lane:1 ],
                    "https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bcl/flowcell_samplesheet.csv",
                    "https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bcl/flowcell.tar.gz"
                    ])
                input[1] = "bclconvert"
                """
                }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.reports,
                    workflow.out.versions,
                    workflow.out.fastq,
                    workflow.out.stats,
                    workflow.out.interop.get(0).get(1).findAll { file(it).name != "IndexMetricsOut.bin" },
                    ).match()
                },
                { assert file(workflow.out.interop.get(0).get(1).find { file(it).name == "IndexMetricsOut.bin" }).exists() }
            )
        }
    }

    test("bcl2fastq") {

        when {

            workflow {
                """
                input[0] = Channel.value([
                    [id:'test', lane:1 ],
                    "https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bcl/flowcell_samplesheet.csv",
                    "https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bcl/flowcell.tar.gz"
                    ])
                input[1] = "bcl2fastq"
                """
                    }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.reports,
                    workflow.out.versions,
                    workflow.out.fastq,
                    workflow.out.stats,
                    workflow.out.interop.get(0).get(1).findAll { file(it).name != "IndexMetricsOut.bin" },
                    ).match()
                },
                { assert file(workflow.out.interop.get(0).get(1).find { file(it).name == "IndexMetricsOut.bin" }).exists() }
            )
        }
    }
}
