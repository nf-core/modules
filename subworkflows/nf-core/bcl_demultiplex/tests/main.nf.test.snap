{
    "bclconvert": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "lane": 1
                    },
                    [
                        "Adapter_Cycle_Metrics.csv:md5,5a0c88793b4a0885fe3dda16609b576e",
                        "Adapter_Metrics.csv:md5,989240b8840b2169ac1061f952c90f6c",
                        "Demultiplex_Stats.csv:md5,93949a8cd96f907d83e0808c1ec2a04b",
                        "Demultiplex_Tile_Stats.csv:md5,83120160b0f22a1303fa1db31c19f6e9",
                        "IndexMetricsOut.bin:md5,9e688c58a5487b8eaf69c9e1005ad0bf",
                        "Index_Hopping_Counts.csv:md5,1059369e375fd8f8423c0f6c934be978",
                        "Quality_Metrics.csv:md5,6614accb1bb414fe312b17b81f5521f7",
                        "Quality_Tile_Metrics.csv:md5,cdc89fd2962bdd4a24f71e186112118a",
                        "RunInfo.xml:md5,03038959f4dd181c86bc97ae71fe270a",
                        "SampleSheet.csv:md5,dc0dffd39541dd6cc5b4801d768a8d2b",
                        "Top_Unknown_Barcodes.csv:md5,2e2faba761137f228e56bd3428453ccc",
                        "fastq_list.csv:md5,05bc84f51840f5754cfb8381b36f2cb0"
                    ]
                ]
            ],
            [
                "versions.yml:md5,a52db0d612f5a3f30d54194d9f7393cf"
            ],
            [
                [
                    {
                        "id": "Sample1_S1_L001",
                        "samplename": "Sample1",
                        "readgroup": {
                            "ID": "000000000-K9H97.1",
                            "PU": "000000000-K9H97.1",
                            "PL": "ILLUMINA",
                            "SM": "Sample1"
                        },
                        "fcid": "test",
                        "lane": 1,
                        "single_end": true
                    },
                    [
                        "Sample1_S1_L001_R1_001.fastq.gz:md5,0675fb6365322eaafb33c0f8e862b54b"
                    ]
                ]
            ],
            [
                
            ],
            [
                "ControlMetricsOut.bin:md5,6d77b38d0793a6e1ce1e85706e488953",
                "CorrectedIntMetricsOut.bin:md5,2bbf84d3be72734addaa2fe794711434",
                "ErrorMetricsOut.bin:md5,38c88def138e9bb832539911affdb286",
                "ExtractionMetricsOut.bin:md5,7497c3178837eea8f09350b5cd252e99",
                "QMetricsOut.bin:md5,7e9f198d53ebdfbb699a5f94cf1ed51c",
                "TileMetricsOut.bin:md5,83891751ec1c91a425a524b476b6ca3c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-07T09:13:44.123805878"
    },
    "bcl2fastq": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "lane": 1
                    },
                    [
                        [
                            [
                                [
                                    [
                                        [
                                            "lane.html:md5,794e48287f47a9f22dcb6b6d0c22c3eb",
                                            "laneBarcode.html:md5,2bbdae3ee57151eab520c966597d7438"
                                        ],
                                        [
                                            
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            "lane.html:md5,f741870307050dcea79a838f5971770f",
                                            "laneBarcode.html:md5,ffe2e863811c76cb9da27d5d124e0611"
                                        ],
                                        [
                                            "lane.html:md5,cca97e771d3491dbc8cd3fe389595b09",
                                            "laneBarcode.html:md5,cca97e771d3491dbc8cd3fe389595b09"
                                        ]
                                    ],
                                    [
                                        [
                                            "lane.html:md5,c383b0768d9978733d3f5d3b91c15af0",
                                            "laneBarcode.html:md5,48842c23b9a2816aec540177df870968"
                                        ],
                                        [
                                            
                                        ]
                                    ]
                                ]
                            ],
                            "Report.css:md5,eb7d3eb68fc1539f411404987246b59b",
                            "index.html:md5,5747c407854ae2c358d0ec201ce622d8",
                            "tree.html:md5,a1b9bf592973ca829ec69ddf888b7e34"
                        ]
                    ]
                ]
            ],
            [
                "versions.yml:md5,513271615cb02dbe541a6202d713ef81"
            ],
            [
                [
                    {
                        "id": "Sample1_S1_L001",
                        "samplename": "Sample1",
                        "readgroup": {
                            "ID": "000000000-K9H97.1",
                            "PU": "000000000-K9H97.1",
                            "PL": "ILLUMINA",
                            "SM": "Sample1"
                        },
                        "fcid": "test",
                        "lane": 1,
                        "single_end": true
                    },
                    [
                        "Sample1_S1_L001_R1_001.fastq.gz:md5,0675fb6365322eaafb33c0f8e862b54b"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "lane": 1
                    },
                    [
                        "AdapterTrimming.txt:md5,48ed2b914b1246c0b5d8667525550946",
                        "ConversionStats.xml:md5,8fe0f57f3f5d256a0762dba75ac62d05",
                        "DemultiplexingStats.xml:md5,2047ff18f5b9107c084de06e9ff943ad",
                        "DemuxSummaryF1L1.txt:md5,03e5fd0c1e3079c5f8c7b4d0501b37ff",
                        "FastqSummaryF1L1.txt:md5,0c6f2d87ee183b84d1051cde9a5643d1",
                        "Stats.json:md5,8e5f038b8aa9e465599d3575f930e604"
                    ]
                ]
            ],
            [
                "ControlMetricsOut.bin:md5,6d77b38d0793a6e1ce1e85706e488953",
                "CorrectedIntMetricsOut.bin:md5,2bbf84d3be72734addaa2fe794711434",
                "ErrorMetricsOut.bin:md5,38c88def138e9bb832539911affdb286",
                "ExtractionMetricsOut.bin:md5,7497c3178837eea8f09350b5cd252e99",
                "QMetricsOut.bin:md5,7e9f198d53ebdfbb699a5f94cf1ed51c",
                "TileMetricsOut.bin:md5,83891751ec1c91a425a524b476b6ca3c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-07T09:01:39.665409003"
    }
}